fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/sorting/FileSorter.java

```diff
@@ -0,0 +1,601 @@
+                return AsyncUtil.READY_FALSE;
+            } else {
+                loadResult = new LoadResult(true, false, mapResult.getSourceContinuation(), mapResult.getSourceNoNextReason());
+                return CompletableFuture.runAsync(() -> saveToNextFile(1), executor).thenApply(vignore -> false);
+            }
+        })).thenApply(vignore -> loadResult);
+    }
+
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    private void saveToNextFile(int maxNumFiles) {
+        final long startTime = System.nanoTime();
+        final boolean compress = adapter.isCompressed();
+        final java.security.Key encryptionKey = adapter.getEncryptionKey();
+        Cipher cipher = null;
+        if (!mapSorter.getMap().isEmpty()) {
+            File file;
+            try {
+                file = adapter.generateFilename();
+                try (FileOutputStream fileStream = new FileOutputStream(file)) {
+                    final FileChannel fileChannel = fileStream.getChannel();
+                    final CodedOutputStream headerStream = CodedOutputStream.newInstance(fileStream);
+                    // To stay the same size, field existence must not change.
+                    final RecordSortingProto.SortFileHeader.Builder fileHeader = RecordSortingProto.SortFileHeader.newBuilder()
+                            .setNumberOfRecords(0)
+                            .setNumberOfSections(0);
+                    headerStream.writeMessageNoTag(fileHeader.build());
+                    final RecordSortingProto.SortSectionHeader.Builder sectionHeader = RecordSortingProto.SortSectionHeader.newBuilder()
+                            .setNumberOfRecords(0)
+                            .setNumberOfBytes(0);
+                    if (encryptionKey != null) {
+                        cipher = CipherPool.borrowCipher();
+                        final byte[] iv = new byte[CipherPool.IV_SIZE];
+                        adapter.getSecureRandom().nextBytes(iv);
+                        cipher.init(Cipher.ENCRYPT_MODE, encryptionKey, new IvParameterSpec(iv));
+                        sectionHeader.setEncryptionIv(ByteString.copyFrom(iv));
+                    }
+                    headerStream.writeMessageNoTag(sectionHeader.build());
+                    final long headerEnd = headerStream.getTotalBytesWritten();
+                    final OutputStream outputStream;
+                    final CodedOutputStream entryStream;
+                    if (compress || cipher != null) {
+                        headerStream.flush();
+                        OutputStream stream = new NoCloseFilterStream(fileStream);
+                        if (cipher != null) {
+                            stream = new CipherOutputStream(stream, cipher);
+                        }
+                        if (compress) {
+                            stream = new DeflaterOutputStream(stream);
+                        }
```

Comment:
It's hard to introduce a strong abstraction, because the coded streams are more stream-like than real streams and even the filter streams used have delicate behavior around when they flush their internal state.

It is possible to pull out the cipher-init and wrap-input/output methods so that they are next to one another and more obviously parallel.